django_kafka/consumer.py

```diff
@@ -91,5 +96,68 @@ def commit_offset(self, msg: cimpl.Message):
             # Explicitly storing offsets after processing gives at-least once semantics.
             self.store_offsets(msg)
 
-    def handle_error(self, error):
+    def retry_msg(self, msg: cimpl.Message, exc: Exception) -> bool:
+        from django_kafka.retry.topic import RetryTopic
+
+        topic = self.get_topic(msg)
+        if not topic.retry:
+            return False
+
+        return RetryTopic(group_id=self.group_id, main_topic=topic).retry_for(
+            msg=msg,
+            exc=exc,
+        )
+
+    def dead_letter_msg(self, msg: cimpl.Message, exc: Exception):
+        from django_kafka.dead_letter.topic import DeadLetterTopic
+
+        topic = self.get_topic(msg)
+        DeadLetterTopic(group_id=self.group_id, main_topic=topic).produce_for(
+            msg=msg,
+            header_message=str(exc),
+            header_detail=traceback.format_exc(),
+        )
+
+    def handle_exception(self, msg: cimpl.Message, exc: Exception):
+        retried = self.retry_msg(msg, exc)
+        if not retried:
+            self.dead_letter_msg(msg, exc)
+            self.log_error(exc)
+
+    def get_topic(self, msg: cimpl.Message) -> "Topic":
+        return self.topics.get_topic(name=msg.topic())
+
+    def log_error(self, error):
         logger.error(error, exc_info=True)
+
+    def process_message(self, msg: cimpl.Message):
+        if msg_error := msg.error():
+            self.log_error(msg_error)
+            return
+
+        try:
+            self.get_topic(msg).consume(msg)
+        # ruff: noqa: BLE001 (we do not want consumer to stop if message consumption fails in any circumstances)
+        except Exception as exc:
+            self.handle_exception(msg, exc)
+
+        self.commit_offset(msg)
+
+    def poll(self) -> Optional[cimpl.Message]:
+        # poll for self.polling_freq seconds
+        return self._consumer.poll(timeout=self.polling_freq)
+
+    def start(self):
+        try:
+            self.subscribe(topics=self.topics.names)
+            while True:
+                if msg := self.poll():
+                    self.process_message(msg)
+        except Exception as exc:
+            self.log_error(exc)
+            raise
+        finally:
+            self.stop()  # commits pending offsets
```

Comment:
Logic: Previously if an exception was not caught outside of [here](https://github.com/RegioHelden/django-kafka/pull/9/files#diff-d637820b6424cb0d21659b35b462d3d0bc7497650c9b67ec22b6e1e0a5075a0dL79), for instance because an exception raised out of `.handle_exception`, then the consumer silently died.

Here we

1. Log the exception, so we know there was an error (sentry log, or in our cases in the console).
2. Called `Consumer.stop()`, so any pending offsets are comitted. 
3. Re-raise, killing the process/consumer. As discussed, any exception raised outside of message consumption should cause a stop in the consumer, because messages are being improperly processed and we need to fix the issue, rather than ignore it and carry on.